### PR TITLE
[DT-1534] [WIP] Feat: PubSub events

### PR DESCRIPTION
**Pull request by DancingAxolotl**
## Changes

- Refactoring Worker repository to add a worker producer for pubsub:
  - Added worker producer that caches domain updates and commits them to PS all at once
  - Moved old worker repo → WorkerDBRepo, removed static functions from WorkerDBRepo
  - Added new WorkerRepo class that extends WorkerDBRepo and conditionally sends data to PS using the new publisher
  - Moved the NamehashLookup logic to producer
- Added Worker Consumer based on worker repo functions that saves data from PS
- Added draft avro schema for pubsub events
- Added new entities to match the avro schema
- Used new entities throughout workers

## To Do

- To get this ready to merge, need to finish [this task](https://linear.app/unstoppable-domains/issue/DT-1556/[rs]-setup-pubsub-migration)
- Can be used as a feature branch for now

## Code Review

This Pull Request was thoroughly reviewed and meets all Code Review Standards pertaining to:

- [ ] **Implementation** - satisfied acceptance criteria
- [ ] **Communication** - notified engineers/stakeholders about impactful changes
- [ ] **Error handling** - handled, logged & alerted on errors
- [ ] **Documentation** - wrote readable code & commits, documented unintuitive code
- [ ] **Testing** - unit & E2E test coverage, tested in staging, DB migrations backwards compatible
- [ ] **Security** - sanitized inputs, vetted dependencies, validated/secured API requests, no committed secrets
- [ ] **Performance** - optimized expensive algorithms & database queries

Please select all **applied** standards relevant to this PR.
## Confirmation

- [ ] **Assignee Confirmation** - I (the author) have filled out the template above
- [ ] **Reviewer Confirmation** - I (a/the reviewer) confirm 1) the author has filled out the template and checked the box that says **Assignee Confirmation** 2) I reviewed the code and selected all applicable items in the code review section.

### CODE REVIEW

File: src/workers/framework/WorkerProducer.ts [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+      messageOrdering: this.config.messageOrderingEnabled,
+      batching: {
+        maxMessages: 100,
+        maxMilliseconds: 1000,
+      },
+    });
+  }
+
+  async isMessageValid(message: object): Promise<boolean> {
+    const errors = await validate(message);
```
Comment (DancingAxolotl) [inline src/workers/framework/WorkerProducer.ts:58]: There is no real validation for now. Need to pick the validation lib and update `Types.ts`

File: src/pubsub/subscribers/BasePubSubSubscriber.ts [hunk 2]
```diff
@@ -12,40 +14,48 @@ export abstract class BasePubSubSubscriber<MessageType> {
...
+        messageId: message.id,
+        publishTime: message.publishTime,
+        received: new Date(message.received),
       };
-    }) as MessageType[];
+    });
   }
 
-  ensureTopicAndSubscriptionExist = async (): Promise<void> => {
+  private async ensureTopicAndSubscriptionExist(): Promise<void> {
```
Comment (tunguyennnnn) [inline src/pubsub/subscribers/BasePubSubSubscriber.ts:53]: feel like this should be `protected`? not necessary for now child class can rewrite the `setup` method and needs to call this method

File: src/workers/framework/schema/domainActionsEvents.avsc.json [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+      "type": "int"
+    },
+    {
+      "name": "blockTimestamp",
+      "type": "int"
+    },
+    {
+      "name": "eventType",
+      "type": "string"
+    },
```
Comment (tunguyennnnn) [inline src/workers/framework/schema/domainActionsEvents.avsc.json:25]: maybe `node` (namehash) always be at this level? It's sort of identity for all domains data right?

Comment (DancingAxolotl) [reply src/workers/framework/schema/domainActionsEvents.avsc.json:29]: Node doesn't appear for some of `reverse resolution` events
You can remove RR only by address for example

File: src/database/migrations/1694098864100-CnsRegistryEventAddBlockTimestamp.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
+import { MigrationInterface, QueryRunner } from 'typeorm';
```
Comment (tunguyennnnn) [inline src/database/migrations/1694098864100-CnsRegistryEventAddBlockTimestamp.ts:1]: lets extract migration and the model into into a PR with the model file to merge?

File: src/env.ts [hunk 5]
```diff
@@ -373,6 +373,7 @@ export const env = {
       'https://api.ud-sandbox.com/metadata',
   },
   PUBSUB: {
+    ENABLE: process.env.PUBSUB_ENABLE || true,
```
Comment (tunguyennnnn) [inline src/env.ts:376]:
```suggestion
    ENABLE: process.env.DOMAIN_ACTIONS_PUBSUB_ENABLE || false,
```
can we wrap this into FEATURE_FLAGS
I also have similar flag for image processing https://github.com/unstoppabledomains/resolution-service/pull/393/files

File: src/env.ts [hunk 6]
```diff
@@ -383,11 +384,17 @@ export const env = {
...
           'onchain.resolution-service.dev.domain-actions.events',
+        SCHEMA: {
+          NAME: 'onchain.resolution-service.dev.domain-actions.schema',
+        },
         SUBSCRIPTIONS: {
           ONCHAIN_DOMAIN_EVENTS_RESOLUTION_SERVICE_SUBSCRIPTION_NAME:
             process.env
               .ONCHAIN_DOMAIN_EVENTS_RESOLUTION_SERVICE_SUBSCRIPTION_NAME ||
             'onchain.resolution-service.domain-actions.dev.events.resolution-apis-subscription',
+          ONCHAIN_DOMAIN_EVENTS_WEBSITE_SUBSCRIPTION_NAME:
```
Comment (tunguyennnnn) [inline src/env.ts:395]: 🤔 why are we including website subscription here?

File: src/pubsub/publishers/PubSubProducer.ts [hunk 7]
```diff
@@ -1,173 +1,151 @@
...
+   * @param eventName
+   * @param message
+   * @param metadata
+   */
+  public async publishEvent(
+    messageContent: PubSubMessage<object>,
+  ): Promise<void> {
+    const message = await this.prepareMessage(messageContent);
+    let publishedMessageId;
+    if (this.schemaType) {
```
Comment (tunguyennnnn) [inline src/pubsub/publishers/PubSubProducer.ts:110]:
```suggestion
    if (!!this.schemaType) {
```
nit

File: src/pubsub/publishers/PubSubProducer.ts [hunk 8]
```diff
@@ -1,173 +1,151 @@
...
+    const messages = await Promise.all(
+      messageContents.map(this.prepareMessage.bind(this)),
+    );
 
     const promises = [];
     for (const message of messages) {
       promises.push(
         (async () => {
-          return this.topic.publishMessage(message);
+          if (this.schemaType) {
```
Comment (tunguyennnnn) [inline src/pubsub/publishers/PubSubProducer.ts:135]:
```suggestion
          if (!!this.schemaType) {
```

File: src/pubsub/publishers/PubSubProducer.ts [hunk 9]
```diff
@@ -1,173 +1,151 @@
...
+import { getPubsub } from '../configs/connect';
+import { PubSubError, PubSubMessage } from '../types';
 
 export abstract class BasePubSubProducer {
   topic: Topic;
   pubsub: PubSub;
   topicId: string;
+  setupDone: boolean;
+  logger: winston.Logger;
+  schemaType: any;
```
Comment (tunguyennnnn) [inline src/pubsub/publishers/PubSubProducer.ts:15]: schema type sounds weird, maybe `schemaEncoder`?

File: src/pubsub/subscribers/BasePubSubSubscriber.ts [hunk 10]
```diff
@@ -12,40 +15,51 @@ export abstract class BasePubSubSubscriber<MessageType> {
   pubsub: PubSub;
   subscription: Subscription;
   v1Subscription: v1.SubscriberClient;
+  logger: winston.Logger;
+  schemaDefinition?: any;
+  schemaType?: any;
```
Comment (tunguyennnnn) [inline src/pubsub/subscribers/BasePubSubSubscriber.ts:20]:
```suggestion
  schemaDecoder?: any;
```

File: src/pubsub/subscribers/BasePubSubSubscriber.ts [hunk 11]
```diff
@@ -12,40 +14,48 @@ export abstract class BasePubSubSubscriber<MessageType> {
...
   }
 
   /**
    *
    * @param message - message
    * @dev - implement this method to handle message in subclass
    */
-  abstract handleMessages(message: MessageType[]): Promise<void>;
+  abstract handleMessages(
+    message: SubscriberData<MessageType>[],
```
[the remark this answers is not in the document]
Comment (tunguyennnnn) [reply src/pubsub/subscribers/BasePubSubSubscriber.ts:36]: 🤔 , I don't think we need to.
In the subscriber we could just pass in
```
DomainActionMesssageType = ResolutionMessageType | ReverseMessageType ...
```

and handle messages by checking with typeof right?

[the remark this answers is not in the document]
Comment (tunguyennnnn) [reply src/pubsub/subscribers/BasePubSubSubscriber.ts:36]: if you concern about decoding the message doesn't know which type then we could let the subclass implement the decoding and control its type.

File: src/pubsub/types.ts [hunk 12]
```diff
@@ -0,0 +1,50 @@
+import {
+  IsNotEmpty,
+  IsOptional,
+  IsString,
+  ValidateNested,
+} from 'class-validator';
+
+import { Attributes } from '../types/common';
+
+export class PubSubError extends Error {}
```
[the remark this answers is not in the document]
Comment (tunguyennnnn) [reply src/pubsub/types.ts:10]: we have a folder to extend errors, src/errors, lets put this there

**Review by tunguyennnnn — COMMENTED** [review]
good start! lets start to extract into smaller PRs to be able to merge?

**Comment by DancingAxolotl** [comment]
Split these changes into smaller PRs:
#441 
#440 
#439 
#438

**Pull request closed without merging**